waspc/cli/src/Wasp/Cli/Command.hs

```diff
@@ -27,3 +43,88 @@ runCommand cmd = do
 -- TODO: What if we want to recognize errors in order to handle them?
 --   Should we add _commandErrorType? Should CommandError be parametrized by it, is that even possible?
 data CommandError = CommandError {_errorTitle :: !String, _errorMsg :: !String}
+
+data Requirement where
+  Requirement :: Requirable r => r -> Requirement
+
+class Typeable r => Requirable r where
+  -- | Check if the requirement is met and return a value representing that
+  -- requirement.
+  --
+  -- This function must always return a value: if the requirement is not met,
+  -- throw a 'CommandError'.
+  checkRequirement :: Command r
+
+-- | Require some dependency or information. See instances of 'Requirable'.
+-- Pattern match on the result of the function to determine the requirement
+-- and possibly get data from the requirement:
+--
+-- @
+-- do
+--   DbConnectionRequirement <- require
+-- @
+require :: Requirable r => Command r
+require =
+  Command (gets (mapMaybe cast)) >>= \case
+    (req : _) -> return req
+    [] -> do
+      -- Requirement hasn't been met, so run the check
+      req <- checkRequirement
+      Command $ modify (Requirement req :)
+      return req
+
+data DbConnectionRequirement = DbConnectionRequirement deriving (Typeable)
+
+instance Requirable DbConnectionRequirement where
+  checkRequirement = do
+    -- NOTE: 'WaspRootRequirement' does not depend on this requirement, so this
+    -- call to 'require' will not result in an infinite loop.
+    WaspRootRequirement waspRoot <- require
+    let outDir = waspRoot SP.</> Cli.Common.dotWaspDirInWaspProjectDir SP.</> Cli.Common.generatedCodeDirInDotWaspDir
+    dbIsRunning <- liftIO $ isDbRunning outDir
+    if dbIsRunning
+      then return DbConnectionRequirement
+      else throwError noDbError
+    where
+      noDbError =
+        CommandError
+          "Can not connect to database"
+          ( "The database needs to be running in order to execute this command."
+              ++ " You can easily start a managed dev database with `wasp start db`."
+          )
+
+-- | Require a Wasp project to exist near the current directory. Get the
+-- project directory by pattern matching on the result of 'require':
+--
+-- @
+-- do
+--   WaspRootRequirement waspRoot <- require
+-- @
+newtype WaspRootRequirement = WaspRootRequirement (SP.Path' SP.Abs (SP.Dir WaspProjectDir)) deriving (Typeable)
```

Comment:
I'm not sure what to name this: when I went through and modified all the places that need to use this, it was split around 50/50 between calling the path `waspRoot` and `waspProjectDir`. We should choose one and stick with it everywhere.

Reply:
Yeah, the issues is that we use both.
I would say WaspProjectDir then.